R/reportDuplicates.R

```diff
@@ -0,0 +1,29 @@
+#' Finds duplicates in a quitte object and warns about them per model and scenario.
+#' Also warn if variables are identical but value or units differ.
+
+#' @param mifdata object that can be converted with as.quitte
+#' @importFrom dplyr filter mutate select
+#' @return only the data that is duplicated. Has 0 rows if everything is fine
+#' @export
+reportDuplicates <- function(mifdata) {
+  mifdata <- as.quitte(mifdata)
+  duplicates <- mifdata %>% 
+    group_by(.data$model, .data$region, .data$variable, .data$period, .data$scenario) %>% 
+    filter(n() > 1) %>%
+    ungroup() %>%
+    droplevels()
+  if (nrow(duplicates) > 0) {
+    short <- duplicates %>%
+      mutate(modelscen = factor(paste(.data$model, .data$scenario)),
+             varunit = factor(paste0(.data$variable, " (", .data$unit, ")"))) %>%
+      select(-c("period", "model", "scenario", "unit", "variable", "region", "value")) %>%
+      unique() %>%
+      droplevels()
+    for (ms in levels(short$modelscen)) {
+      shortms <- droplevels(filter(short, .data$modelscen %in% ms))
+      warning("Duplicated data found for ", ms, ":\n",
+              "Variables: ", paste(levels(shortms$varunit), collapse = ", "))
+    }
+  }
+  return(duplicates)
```

Comment:
Couple of points:
1. This potentially returns a lot of warnings, where one warning would suffice:
   ```
     if (nrow(duplicates) > 0) {
       warning(paste(c(
         "Duplicated data found for:",
         duplicates %>%
           distinct(model, scenario, variable, unit) %>%
           group_by(.data$model, .data$scenario) %>%
           summarise(
             text = paste0(
               unique(.data$model), ' ', unique(.data$scenario), ': ',
               paste(paste0(.data$variable, '(', .data$unit, ')'),
                     collapse = ', ')),
             .groups = 'drop') %>%
           pull('text')),
         collapse = '\n  '))
     }
   ```
   ```
   reportDuplicates(bind_rows(quitte_example_data, quitte_example_data))
   Warning message:
   In reportDuplicates(bind_rows(quitte_example_data, quitte_example_data)) :
     Duplicated data found for:
     REMIND r7552c_1p5C_Def-rem-5: Consumption(billion US$2005/yr), GDP|PPP(billion US$2005/yr), Population(million), PE|Coal(EJ/yr), PE|Oil(EJ/yr), PE|Gas(EJ/yr), PE|Biomass(EJ/yr), PE|Nuclear(EJ/yr), PE|Hydro(EJ/yr), PE|Wind(EJ/yr), PE|Solar(EJ/yr), PE|Geothermal(EJ/yr), PE(EJ/yr), SE|Heat|Coal(EJ/yr), SE|Heat|Gas(EJ/yr), Emi|CO2(Mt CO2/yr), Policy Cost|Consumption Loss(billion US$2005/yr)
     REMIND r7552c_1p5C_UBA_Sust-rem-5: Consumption(billion US$2005/yr), GDP|PPP(billion US$2005/yr), Population(million), PE|Coal(EJ/yr), PE|Oil(EJ/yr), PE|Gas(EJ/yr), PE|Biomass(EJ/yr), PE|Nuclear(EJ/yr), PE|Hydro(EJ/yr), PE|Wind(EJ/yr), PE|Solar(EJ/yr), PE|Geothermal(EJ/yr), PE(EJ/yr), SE|Heat|Coal(EJ/yr), SE|Heat|Gas(EJ/yr), Emi|CO2(Mt CO2/yr), Policy Cost|Consumption Loss(billion US$2005/yr)
     REMIND r7552c_2C_Def-rem-5: Consumption(billion US$2005/yr), GDP|PPP(billion US$2005/yr), Population(million), PE|Coal(EJ/yr), PE|Oil(EJ/yr), PE|Gas(EJ/yr), PE|Biomass(EJ/yr), PE|Nuc [... truncated]
   ```
2. This has effectively two return paths: the warning (or lack thereof) and the returned data.  
   - In batch use, the warning is more of a distraction.  It would need to be suppressed and captured in order to check, and one can more easily check the return value.
   - In interactive use, people might be more interested in the warning, to know whether the data contains duplicates, and only want to know what those duplicates are in step two.

   I would probably factor out some `has_duplicates()` function for interactive use, and ditch the warning entirely from `reportDuplicates()`.  Or add an option to either suppress the warning or to show the warning, but return the duplicates invisible().